File: API/CSharp_API/Executors/Executor.cs
using Framework.Exceptions;
using System.Reflection;

namespace Framework;

public abstract class Executor
{
	public virtual void Init(ExecutorSettings settings)
	{

	}

	public virtual void DoPost(HttpRequest request, HttpResponse response)
	{
		throw new NotImplementedDoMethodException(nameof(DoPost), GetType());
	}

	public virtual void DoGet(HttpRequest request, HttpResponse response)
	{
		throw new NotImplementedDoMethodException(nameof(DoGet), GetType());
	}

	public virtual void DoHead(HttpRequest request, HttpResponse response)
	{
		throw new NotImplementedDoMethodException(nameof(DoHead), GetType());
	}

	public virtual void DoPut(HttpRequest request, HttpResponse response)
	{
		throw new NotImplementedDoMethodException(nameof(DoPut), GetType());
	}

	public virtual void DoDelete(HttpRequest request, HttpResponse response)
	{
		throw new NotImplementedDoMethodException(nameof(DoDelete), GetType());
	}

	public virtual void DoPatch(HttpRequest request, HttpResponse response)
	{
		throw new NotImplementedDoMethodException(nameof(DoPatch), GetType());
	}

	public virtual void DoOptions(HttpRequest request, HttpResponse response)
	{
#if DEBUG
		List<string> methods = ["OPTIONS"];
		Type type = GetType();
		string allowHeaders = "";

		IsImplemented(methods, type, "POST", nameof(DoPost));
		IsImplemented(methods, type, "GET", nameof(DoGet));
		IsImplemented(methods, type, "HEAD", nameof(DoHead));
		IsImplemented(methods, type, "PUT", nameof(DoPut));
		IsImplemented(methods, type, "DELETE", nameof(DoDelete));
		IsImplemented(methods, type, "PATCH", nameof(DoPatch));
		IsImplemented(methods, type, "TRACE", nameof(DoTrace));
		IsImplemented(methods, type, "CONNECT", nameof(DoConnect));

		for (int i = 0; i < methods.Count; i++)
		{
			allowHeaders += methods[i];

			if (i + 1 != methods.Count)
			{
				allowHeaders += ", ";
			}
		}

		response.AddHeader("Allow", allowHeaders);
#else
		throw new NotImplementedDoMethodException(nameof(DoOptions), GetType());
#endif
	}

	public virtual void DoTrace(HttpRequest request, HttpResponse response)
	{
#if DEBUG
		response.SetBody(request.GetHttpRawRequest());
#else
		throw new NotImplementedDoMethodException(nameof(DoTrace), GetType());
#endif
	}

	public virtual void DoConnect(HttpRequest request, HttpResponse response)
	{
		throw new NotImplementedDoMethodException(nameof(DoConnect), GetType());
	}

	public abstract Utility.ExecutorType GetExecutorType();

	public abstract void Destroy();

	private static void IsImplemented(List<string> methods, Type type, string requestName, string methodName)
	{
		MethodInfo? method = type.GetMethod(methodName, BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);

		if (method == null)
		{
			return;
		}

		if (method.GetBaseDefinition().DeclaringType != method.DeclaringType)
		{
			methods.Add(requestName);
		}
	}
}

File: WebFrameworkExecutors/src/Executors/Executor.cpp
#include "Executors/Executor.h"

#include <Strings.h>

static void isImplemented
(
	std::vector<std::string>& result,
	framework::interfaces::IHttpRequest& request, 
	framework::interfaces::IHttpResponse& response,
	const std::string& methodName, 
	void(framework::Executor::*method)(framework::interfaces::IHttpRequest&, framework::interfaces::IHttpResponse&), 
	framework::Executor& executor
);

namespace framework
{
	void (Executor::* Executor::getMethod(std::string_view methodName))(interfaces::IHttpRequest&, interfaces::IHttpResponse&)
	{
		static const ::utility::strings::string_based_unordered_map<void(Executor::*)(interfaces::IHttpRequest&, interfaces::IHttpResponse&)> methods =
		{
			{ "GET", &Executor::doGet },
			{ "POST", &Executor::doPost },
			{ "HEAD", &Executor::doHead },
			{ "PUT", &Executor::doPut },
			{ "DELETE", &Executor::doDelete },
			{ "PATCH", &Executor::doPatch },
			{ "OPTIONS",&Executor::doOptions },
			{ "TRACE", &Executor::doTrace },
			{ "CONNECT", &Executor::doConnect }
		};

		return methods.find(methodName)->second;
	}

	void Executor::init(const utility::JSONSettingsParser::ExecutorSettings& settings)
	{

	}

	void Executor::doPost(interfaces::IHttpRequest& request, interfaces::IHttpResponse& response)
	{
		throw exceptions::NotImplementedException(typeid(*this).name(), __func__);
	}

	void Executor::doGet(interfaces::IHttpRequest& request, interfaces::IHttpResponse& response)
	{
		throw exceptions::NotImplementedException(typeid(*this).name(), __func__);
	}

	void Executor::doHead(interfaces::IHttpRequest& request, interfaces::IHttpResponse& response)
	{
		throw exceptions::NotImplementedException(typeid(*this).name(), __func__);
	}

	void Executor::doPut(interfaces::IHttpRequest& request, interfaces::IHttpResponse& response)
	{
		throw exceptions::NotImplementedException(typeid(*this).name(), __func__);
	}

	void Executor::doDelete(interfaces::IHttpRequest& request, interfaces::IHttpResponse& response)
	{
		throw exceptions::NotImplementedException(typeid(*this).name(), __func__);
	}

	void Executor::doPatch(interfaces::IHttpRequest& request, interfaces::IHttpResponse& response)
	{
		throw exceptions::NotImplementedException(typeid(*this).name(), __func__);
	}

	void Executor::doOptions(interfaces::IHttpRequest& request, interfaces::IHttpResponse& response)
	{
#ifdef NDEBUG
		throw exceptions::NotImplementedException(typeid(*this).name(), __func__);
#endif

		std::vector<std::string> methods = { "OPTIONS" };
		std::string allowHeader;

		isImplemented(methods, request, response, "GET", &Executor::doGet, *this);
		isImplemented(methods, request, response, "POST", &Executor::doPost, *this);
		isImplemented(methods, request, response, "HEAD", &Executor::doHead, *this);
		isImplemented(methods, request, response, "PUT", &Executor::doPut, *this);
		isImplemented(methods, request, response, "DELETE", &Executor::doDelete, *this);
		isImplemented(methods, request, response, "PATCH", &Executor::doPatch, *this);
		isImplemented(methods, request, response, "TRACE", &Executor::doTrace, *this);
		isImplemented(methods, request, response, "CONNECT", &Executor::doConnect, *this);

		for (size_t i = 0; i < methods.size(); i++) 
		{
			allowHeader += methods[i];

			if (i + 1 != methods.size()) 
			{
				allowHeader += ", ";
			}
		}

		response.addHeader("Allow", allowHeader.data());
	}

	void Executor::doTrace(interfaces::IHttpRequest& request, interfaces::IHttpResponse& response)
	{
#ifdef NDEBUG
		throw exceptions::NotImplementedException(typeid(*this).name(), __func__);
#endif

		response.setBody(request.getRawRequest());
	}

	void Executor::doConnect(interfaces::IHttpRequest& request, interfaces::IHttpResponse& response)
	{
		throw exceptions::NotImplementedException(typeid(*this).name(), __func__);
	}
}

void isImplemented
(
	std::vector<std::string>& result,
	framework::interfaces::IHttpRequest& request,
	framework::interfaces::IHttpResponse& response,
	const std::string& methodName, 
	void(framework::Executor::*method)(framework::interfaces::IHttpRequest&, framework::interfaces::IHttpResponse&), 
	framework::Executor& executor
)
{
	response.setDefault();

	try
	{
		(executor.*method)(request, response);

		result.push_back(methodName);
	}
	catch (const framework::exceptions::NotImplementedException&)
	{

	}
	catch (...)
	{
		result.push_back(methodName);
	}
}
